Replace pyjwt with authlib in `org.matrix.login.jwt`

aytchell opened the pull request:
As discussed in #12830 here is the PR for replacing PyJWT with Authlib

### Contained changes

Beneath the actual code changes I have
  * adapted the documentation in `jwt.md`
  * changed comments pointing to the list of supported algorithms
  * added a small dev-script for creating JWTs locally (which seems to be included in pyjwt but not in authlib)
  * Changed the expected error messages in some unit tests
  * removed pyjwt from `pyproject.toml` and ran `poetry lock --no-update`

### Pull Request Checklist

<!-- Please read https://matrix-org.github.io/synapse/latest/development/contributing_guide.html before submitting your pull request -->

* [X] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [X] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

aytchell commented:
I've ssen that tests are failing because `pyproject.toml` and `poetry.lock` are out of sync.

poetry is kind of brand new to me and I don't want to fiddle around with dependency versions of a big project. So I reverted my change to `pyproject.toml` and leave it to someone more experience to remove the pyjwt entry in there.

aytchell commented:
Ugh, sorry for all the red test runs. I'm currently figuring out the behaviour of the build bot.
The test `Tests / trial (3.10) (pull_request)` fails with `builtins.ModuleNotFoundError: No module named 'authlib'`

Does it not do `poetry install --extras all`? (and if not - are there no tests for oidc?)
Or do I have to enable this somehow in login_tests?

DMRobertson commented on synapse/rest/client/login.py:
```diff
@@ -420,25 +422,51 @@ async def _do_jwt_login(
             )
 
-        import jwt
+        jwt = JsonWebToken([self.jwt_algorithm])
```
I think you'll need to add the imports from authlib (lines +29, +30) here.

aytchell replied:
Thanks for this hint. I'll move the imports and run tests for "all extras" and "no extras"

DMRobertson commented:
> Does it not do `poetry install --extras all`

It does for most jobs, but not all: because some people will be running Synapse without optional dependencies like authlib. It's still marked as optional in pyproject.toml:

https://github.com/matrix-org/synapse/blob/00d915b2a73c8f0dea43d715bd98f7a736167062/pyproject.toml#L168

DMRobertson commented:
> I've seen that tests are failing because `pyproject.toml` and `poetry.lock` are out of sync.

Got a link? I think there's a bug in the beta version of poetry that we use where it can spuriously give out these warning messages in CI; I think this might be a false positive.

aytchell commented:
> > I've seen that tests are failing because pyproject.toml and poetry.lock are out of sync.

> Got a link? I think there's a bug in the beta version of poetry that we use where it can spuriously give out
> these warning messages in CI; I think this might be a false positive.

I was talking about the first version of my PR.

I removed `pyjwt` from `pyproject.toml` and did `poetry lock` which then caused poetry to "go wild" and update versions from *a lot* of other libs. So I did commit the change in `pyproject.toml` but not the changed `poetry.lock` which then was disliked by the build bot.

So long story short: I removed all code which makes use of pyjwt but did not remove the dependency entries in `pyproject.toml` nor `poetry.lock`

DMRobertson commented:
I'm pretty sure the remaining test failures is another known flake.

DMRobertson commented on docs/jwt.md:
```diff
@@ -36,28 +36,15 @@ As with other login types, there are additional fields (e.g. `device_id` and
 
 ## Preparing Synapse
 
-The JSON Web Token integration in Synapse uses the
-[`PyJWT`](https://pypi.org/project/pyjwt/) library, which must be installed
-as follows:
-
- * The relevant libraries are included in the Docker images and Debian packages
-   provided by `matrix.org` so no further action is needed.
-
- * If you installed Synapse into a virtualenv, run `/path/to/env/bin/pip
-   install synapse[pyjwt]` to install the necessary dependencies.
-
- * For other installation mechanisms, see the documentation provided by the
-   maintainer.
-
-To enable the JSON web token integration, you should then add an `jwt_config` section
```
As discussed below, server owners will now need the `authlib` package instead of `pyjwt` to support authlib.

Can you restore this and make the following changes:
- refer to `authlib` rather than `PyJWT`
- change `pip install synapse[pyjwt]` to `pip install synapse[jwt]` (as it should have been originally!)
  - Note that the thing in square brackets is the name of the extra feature, not the name of a package.

aytchell replied:
Just saw, that the changed URL in `config/jwt.py` collides with #12941 so I merged and fixed this locally.

Then I saw that each PR has _one_ commit ... so you're squashing before a merge ... which means I'll rebase all my commits.

Is that correct?

DMRobertson replied:
> Then I saw that each PR has one commit ... so you're squashing before a merge ... which means I'll rebase all my commits.
>
> Is that correct?

Assuming you merged in the `develop` branch, you should be okay to push the merge commit to this branch. (Where possible we prefer to avoid rebases on PRs since it makes reviewing more difficult.)

DMRobertson commented on scripts-dev/build_custom_jwt.py:
```diff
@@ -0,0 +1,58 @@
+    print(create_HS256_jwt(payload, secret))
+
+
+if __name__ == "__main__":
```
From what you've written in the docs above, this is meant to be a substitution for pyjwt's commandline functionality. If we want to include this as a script I think I'd expect this to take command line arguments using e.g. `argparse`, rather than have people edit this script in-place.

Would a short, copy-and-pastable code sample in the docs perhaps be more appropriate?

aytchell replied:
I confess that it's a somehow "half-baked developer's tools". The current documentation mentions a command line tool from pyjwt ... which now is right out.

So then to keep things simple and not clutter the scripts directory with stuff from extra features I'd like to take the "code snippet in docs" option.

DMRobertson reviewed (changes requested):
Many thanks for this. I don't follow all the details, so what follows is a mixture of questions and requested changes.

aytchell commented on docs/sample_config.yaml:
```diff
@@ -2184,7 +2184,7 @@ sso:
     #
     # Supported algorithms are listed at
-    # https://pyjwt.readthedocs.io/en/latest/algorithms.html
+    # https://docs.authlib.org/en/latest/specs/rfc7518.html
```
Good point. Afaik the "normal" JWT issued by an oidc provider is a JWS (signed) and authlib also supports JWE (encrypted); JWK as far I understood [RFC 7518](https://datatracker.ietf.org/doc/html/rfc7518#section-6) is not a JWT.
I think that in the context of authentication a JWE doesn't make sense (see "Overview" in [RFC 7519](https://www.rfc-editor.org/rfc/rfc7519#section-3) )

I'll add a brief note to the "two URL locations".

aytchell commented on docs/usage/configuration/config_documentation.md:
```diff
@@ -2947,7 +2947,7 @@ Additional sub-options for this setting include:
    decode the contents of the JSON web token. Required if `enabled` is set to true.
 * `algorithm`: The algorithm used to sign the JSON web token. Supported algorithms are listed at
-   https://pyjwt.readthedocs.io/en/latest/algorithms.html Required if `enabled` is set to true.
+   https://docs.authlib.org/en/latest/specs/rfc7518.html Required if `enabled` is set to true.
```
yep, also added 'section JWS' to the link text

aytchell commented on pyproject.toml:
```diff
@@ -196,7 +195,6 @@ systemd = ["systemd-python"]
 url_preview = ["lxml"]
 sentry = ["sentry-sdk"]
 opentracing = ["jaeger-client", "opentracing"]
-jwt = ["pyjwt"]
```
You're correct. I'll re-add this line and update `poetry.lock`

aytchell commented on pyproject.toml:
```diff
@@ -230,8 +228,6 @@ all = [
     "sentry-sdk",
     # opentracing
     "jaeger-client", "opentracing",
-    # jwt
-    "pyjwt",
```
extended comment and regenerated lockfile

aytchell commented on synapse/rest/client/login.py:
```diff
@@ -420,25 +420,54 @@ async def _do_jwt_login(
+        try:
+            claims.validate(leeway=120)  # allows 2 min of clock skew
+
+            # Enforce the old behavior which is rolled out in productive
+            # servers: if the JWT contains an 'aud' claim but none is
+            # configured, the login attempt will fail
+            if claims.get("aud") is not None:
+                if self.jwt_audiences is None or len(self.jwt_audiences) == 0:
+                    raise InvalidClaimError("aud")
```
The current implementation of the JWT login flow mostly uses the semantics of OAuth2 (not OIDC which adds an extra layer).

But here's a difference in the details: According to OAuth2 the "aud" claim is optional and if there's one given in the JWT but the resource server (i.e. synapse) is not interested in it ... that's OK (so authlib doesn't raise an exception in this case)

In contrast the comment in the `jwt_config` states
```
    # Note that if the "aud" claim is included in a JSON web token then
    # validation will fail without configuring audiences.
```
(and there is a unit test for this statement).

That's what these therre lines do.

DMRobertson replied:
Let me see if I can summarise. Is this accurate?

- OAuth2 and hence Authlib are happy for an `aud` claim to exist in JWTs but be ignored by the resource server. 
- Synapse can be configured with a list of audiences.
- If so, we tell authlib above (line 431) that the audience claim is required in JWTs and should validate against the 
configured list of audiences.
  - Authlib will verify the aud claim matches the configured list of audiences
  - additionally, `essential=True` means that Authlib will raise if the `aud` claim is missing.
- **If Synapse is _not_ configured with a list of audiences, the existing behaviour is to reject JWTs that include an aud claim.**
    - we want to preserve this behaviour.

DMRobertson replied:
I'm just wondering if there is some way we can get authlib to enforce the bolded bullet for us. I couldn't see any obvious way to do so, looking at how `JWTClaims` and `BaseClaims` works. (Maybe a subclass of `JWTClaims` that overrides `validate_aud`? But that'd be overkill.)

In any case, LGTM.

aytchell commented on poetry.lock:
```diff
@@ -815,7 +815,7 @@ python-versions = ">=3.5"
 name = "pyjwt"
 version = "2.4.0"
 description = "JSON Web Token implementation in Python"
-category = "main"
+category = "dev"
 optional = false
 python-versions = ">=3.6"
 
```
`poetry lock --no-update` did this. I was simply glad that it didn't alter all the versions and kept this :-)

DMRobertson commented on synapse/rest/client/login.py:
```diff
@@ -420,25 +422,51 @@ async def _do_jwt_login(
                 403,
                 "JWT validation failed: %s" % (str(e),),
                 errcode=Codes.FORBIDDEN,
             )
```
Thanks, that makes sense. Let's keep it but leave a brief comment explaining what you just said.

DMRobertson commented on synapse/rest/client/login.py:
```diff
@@ -420,25 +420,54 @@ async def _do_jwt_login(
+                claims_cls=JWTClaims,
+                claims_options=claim_options,
+            )
+        except BadSignatureError:
```
```suggestion
        except BadSignatureError:
            # We handle this case separately to provide a better error message,
```

DMRobertson commented on synapse/rest/client/login.py:
```diff
@@ -420,25 +420,54 @@ async def _do_jwt_login(
+        from authlib.jose.errors import BadSignatureError, InvalidClaimError, JoseError
+
+        jwt = JsonWebToken([self.jwt_algorithm])
+        claim_options = {}
+        if self.jwt_issuer is not None:
+            claim_options["iss"] = {"value": self.jwt_issuer, "essential": True}
+        if self.jwt_audiences is not None:
+            claim_options["aud"] = {"values": self.jwt_audiences, "essential": True}
```
Thanks, this is very clear!

DMRobertson commented on synapse/rest/client/login.py:
```diff
@@ -420,25 +420,54 @@ async def _do_jwt_login(
             )
 
-        user = payload.get(self.jwt_subject_claim, None)
+        try:
+            claims.validate(leeway=120)  # allows 2 min of clock skew
+
+            # Enforce the old behavior which is rolled out in productive
+            # servers: if the JWT contains an 'aud' claim but none is
+            # configured, the login attempt will fail
+            if claims.get("aud") is not None:
+                if self.jwt_audiences is None or len(self.jwt_audiences) == 0:
+                    raise InvalidClaimError("aud")
+        except JoseError as e:
+            raise LoginError(
+                403,
+                "JWT validation failed: %s" % (str(e),),
+                errcode=Codes.FORBIDDEN,
+            )
```
Oh yeah, this is from line -433 to -439.

DMRobertson commented on synapse/config/jwt.py:
```diff
@@ -41,13 +35,6 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
             self.jwt_issuer = jwt_config.get("issuer")
             self.jwt_audiences = jwt_config.get("audiences")
-
-            try:
-                import jwt
-
-                jwt  # To stop unused lint.
-            except ImportError:
-                raise ConfigError(MISSING_JWT)
```
I think this is still outstanding.

DMRobertson reviewed (changes requested):
Thanks, nearly there!

aytchell commented:
> I think this is still outstanding.

Correct. I've already added this to my local branch; it now reads `from authlib.jose import JsonWebToken` and the error message states to install it with `pip install synapse[jwt]` (as mentioned in the documentation).

I've prepared a version which is merged with the latest develop as well as a version which is rebased atop of the latest develop.

Which is the preferred one to procede with?

DMRobertson commented:
> I've prepared a version which is merged with the latest develop as well as a version which is rebased atop of the latest develop.
> 
> Which is the preferred one to procede with?

I mean to cover this [here](https://github.com/matrix-org/synapse/pull/13011#discussion_r897817518), but to be explicit: the one merged with develop please!

aytchell commented:
Oh. Thanks. I missed this comment.
Pushed it

DMRobertson commented on synapse/config/jwt.py:
```diff
@@ -43,11 +43,11 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
-                import jwt
+                from authlib.jose import JsonWebToken
 
-                jwt  # To stop unused lint.
+                JsonWebToken  # To stop unused lint.
             except ImportError:
-                raise ConfigError(MISSING_JWT)
+                raise ConfigError(MISSING_AUTHLIB)
```
I'd prefer this to be a  call to `check_dependencies` (which checks the version of authlib installed). Additionally this shouldn't raise a ConfigError: the config may be perfectly valid; it's your installation that's screwed up.

However there's already a lot going on here, and there are other placed with the same pattern. I'll raise a new issue for this.

DMRobertson replied:
That's fair. #13066, for completeness

DMRobertson reviewed (approved):
Thanks, this LGTM! I'm going to ask for a quick second opinion from the Synapse team though, just because I'm not massively familiar with OAuth2.

aytchell commented:
Thanks a lot for guiding me through this process

DMRobertson commented on synapse/rest/client/login.py:
```diff
@@ -420,25 +420,55 @@ async def _do_jwt_login(
+                errcode=Codes.FORBIDDEN,
             )
-        except jwt.PyJWTError as e:
+        except Exception as e:
```
(That will mean that any other exception gets properly flagged as an application error rather than causing a 403 to the requester.)

DMRobertson reviewed (changes requested):
Sorry, one final change and this is ready to go

DMRobertson reviewed (approved):
Thank you so much!

DMRobertson commented:
I think this conflicts with Erik's recent change to `poetry.lock` in #13063. I'll fixup and merge
